[SNAP FORK][UPSTREAMED] enable support multiplex workers in busybox actions

mauriciogg opened the pull request:
Conditionally enable worker and multiples support for resource processing actions using flags
 `--define=persistent_android_resource_processor`
 `--define=persistent_multiplex_android_resource_processor` -> this flag will implicitly enable the `persistent_android_resource_processor`

 - Disable warnings for busybox actions Warnings go to stdout/stderr which makes it incompatible with workers
 - Force use of param files



fixes: https://github.com/bazelbuild/rules_android/issues/30

ted-xie commented on rules/busybox.bzl:
```diff
@@ -118,6 +118,10 @@ def _make_resources_flag(
 
+def _disable_warnings(args):
+    # Disable warnings - this are output to stdin/stderr which breaks worker mode
+    args.add("--logWarnings=false")
```
What happens to warnings in worker mode? I think we'd still want to know about them. Would they get dumped to a non-stdout/stderr file?

Bencodes replied:
It looks like this is just mirroring the existing behavior inside `BusyBoxActionBuilder` https://github.com/bazelbuild/bazel/blob/84c1ed430405b154b6e9eb2c28281f450e250eff/src/main/java/com/google/devtools/build/lib/rules/android/BusyBoxActionBuilder.java#L369

ted-xie reviewed (changes requested):
TL;DR we need to be able to optionally enable/disable workers and multiplex workers. It would also be ideal to add a test for this, but we haven't open-sourced the tests yet, so it's acceptable to omit for now.

mauriciogg commented on rules/busybox.bzl:
```diff
@@ -388,16 +393,20 @@ def _package(
         outputs = output_files,
         mnemonic = "PackageAndroidResources",
         progress_message = "Packaging Android Resources in %s" % ctx.label,
+        supports_workers = True,
```
Done. Default nos is false. Can be enabled using `--define=persistent_android_resource_processor=true`

ted-xie replied:
It looks to me like the `supports_workers` and `supports_multiplex_workers` arguments to `java_run()` no longer do anything. Maybe we can remove these from the PR?

mauriciogg replied:
Done.

mauriciogg commented on rules/busybox.bzl:
```diff
@@ -388,16 +393,20 @@ def _package(
     if java_package:
         args.add("--packageForR", java_package)
 
+    _disable_warnings(args)
```
Done. Default nos is false. Can be enabled using `--define=persistent_android_resource_processor=true`

mauriciogg commented on rules/busybox.bzl:
```diff
@@ -388,16 +393,20 @@ def _package(
         mnemonic = "PackageAndroidResources",
         progress_message = "Packaging Android Resources in %s" % ctx.label,
+        supports_workers = True,
+        supports_multiplex_workers = True,
```
Done

ted-xie commented:
Would you mind rebasing this PR? There are merge conflicts with pre-alpha HEAD now.

ted-xie commented on rules/busybox.bzl:
```diff
@@ -118,6 +119,12 @@ def _make_resources_flag(
     )
 
+def _disable_warnings(ctx, args):
+    if (_flags.get(ctx).persistent_android_resource_processor or
```
It turns out that we have some internal test rules that directly call into the busybox. It was a pain to manually add `_flags = [...]` to each of those rules, so I decided to add an early-fail here:

```
if not hasattr(ctx, "_flags"):
  return
```

ted-xie commented on rules/busybox.bzl:
```diff
@@ -759,18 +793,30 @@ def _merge_compiled(
     )
 
+def _java_run(ctx, *args, **kwargs):
+    enable_workers = _flags.get(ctx).persistent_android_resource_processor
+    multiplex_workers = _flags.get(ctx).persistent_multiplex_android_resource_processor
+    kwargs["supports_workers"] = enable_workers or multiplex_workers
+    kwargs["supports_multiplex_workers"] = multiplex_workers 
```
I put these in an if-block:

```
enable_workers = False
multiplex_workers = False
if hasattr(ctx, "_flags"):
  enable_workers = [...]
  multiplex_workers = [...]
  kwargs[...]
  kwargs[...]
```

ted-xie reviewed (changes requested):
LMK if my proposed changes make sense to you. I'll also discuss this more internally.

mauriciogg commented:
> LMK if my proposed changes make sense to you. I'll also discuss this more internally.

makes sense. I deleted the use of the flags and replaced with `ctx.fragments.android.persistent_android_resource_processor` and `ctx.fragments.android.persistent_multiplex_android_resource_processor`

maybe it makes sense to replace this logic and add the starlark actions to the expansion in https://github.com/bazelbuild/bazel/blob/56fd3dc889717a8428843137e18016c906129924/src/main/java/com/google/devtools/build/lib/rules/android/AndroidConfiguration.java#L863

ted-xie commented:
The updated changes look good.

At a high level, we'd like to move as much Android-related stuff out of bazel as possible (which is really one of the major motivations behind having this separate Android rules codebase), so I think it's better for now to keep all the logic inside rules_android. @ahumesky, thoughts?

ted-xie commented:
I'm actually getting this error when integrating into google3:

```
File "rules/busybox.bzl", line 122, column 30, in _disable_warnings
                if (ctx.fragments.android.persistent_android_resource_processor or
Error: 'android' value has no field or method 'persistent_android_resource_processor'
```

@mauriciogg Have you tested this change out locally to see if it works with Bazel? I don't see persistent_android_resource_processor in AndroidConfigurationApi.java, which is where flags are publicly exposed and ultimately hooked up to ctx.fragments. Example [here](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/starlarkbuildapi/android/AndroidConfigurationApi.java;l=241-246?q=f:AndroidConfigurationApi.java).

I think to merge this change, we'll first have to expose persistent_android_resource_processor to AndroidConfigurationApi, wait for that change to propagate to both a Bazel prerelease and a Blaze release, and only then will this cleanly integrate with internal code.

ted-xie commented on rules/busybox.bzl:
```diff
@@ -118,6 +118,12 @@ def _make_resources_flag(
     )
 
+def _disable_warnings(ctx, args):
+    if (ctx.fragments.android.persistent_android_resource_processor or
```
I changed these fragment flags to `ctx.fragments.android.persistent_busybox_tools` and `ctx.fragments.android.persistent_multiplex_busybox_tools` internally and tests are passing. Those flags also seem more appropriate than the resource_processor ones; what do you think?

Unfortunately the `fragments.android` data structure can't accept `Void` expansion flags such as persistent_android_resource_processor, and changing the return type of that flag handler to `boolean` causes some other kind of sanity check error. I think in Bazel/Blaze, expansion flags must return Void, so there's somewhat of an impedance mismatch.

ted-xie commented:
This has been merged.

nkoroste commented:
Looks like this was reverted https://github.com/bazelbuild/rules_android/commit/f49f2e3af10a32f813ca294a0f025fd27294861a

ted-xie commented:
Yes - the root cause was that in google3, there can be spaces in [manifest_values](https://github.com/bazelbuild/rules_android/blob/20d9b41ea90bc84b649c90663863480a03dd13ed/rules/android_binary_internal/attrs.bzl#L69), i.e. `manifest_values = {"versionName": "a b c d"}`. There was an issue with this PR where those spaces were not properly quoted, which was causing internal app builds to fail. If you sed all of the `multiline` usages in busybox.bzl and replace with `shell` instead (see doc [here](https://bazel.build/rules/lib/builtins/Args#set_param_file_format)), it should fix the problem. I'm trying that out internally and wil attempt to un-revert this change.

ThomasCJY commented:
@ted-xie Did you get the chance to try whether `replacing multiline usages in busybox.bzl and with shell` fix the issue? If so, can we re-introduce this change?

ThomasCJY commented:
nvm, I saw it has been re-introduced in https://github.com/bazelbuild/rules_android/commit/49700f23ee7e931522cbd26908288e5158a5d813
